File: 2021/Day9/Day9.scala
import scala.io.Source
import scala.annotation.tailrec

@main def day9 =
    type Point = (Int, Int)
    type Ocean = Vector[Vector[Int]]

    extension (ocean: Ocean) def |=>(cell: Point) = ocean(cell._2)(cell._1)

    def neighbors(X: Int, Y: Int)(cell: Point) =
        def move(from: Point)(dir: Point) =
            val (x, y) = from
            val (vx, vy) = dir
            (x + vx, y + vy)
        def inBound(cell: Point) =
            val (x, y) = cell
            !(x < 0 || x > X || y < 0 || y > Y)

        Vector((0, 1), (0, -1), (1, 0), (-1, 0))
            .map(move(cell))
            .filter(inBound)

    def isLow(ocean: Ocean)(cell: Point) =
        val X = ocean(0).length - 1
        val Y = ocean.length - 1
        val current = ocean |=> cell

        neighbors(X, Y)(cell)
            .forall(cell => (ocean |=> cell) > current)

    def lowest(ocean: Ocean) =
        val X = ocean(0).length - 1
        val Y = ocean.length - 1

        (for x <- 0 to X
        yield for y <- 0 to Y
        yield (x, y))
            .flatMap(identity)
            .filter(isLow(ocean))
            .toVector

    def firstPart(ocean: Ocean) =
        lowest(ocean)
            .map(cell => (ocean |=> cell) + 1)
            .sum

    def basin(ocean: Ocean)(cell: Point) =
        val X = ocean(0).length - 1
        val Y = ocean.length - 1

        def flood(visited: Set[Point], cell: Point): Set[Point] =
            if visited.contains(cell) || (ocean |=> cell) == 9 then visited
            else
                neighbors(X, Y)(cell)
                    .foldLeft(visited + cell)(flood)

        flood(Set(), cell).size

    def secondPart(ocean: Ocean) =
        lowest(ocean)
            .map(basin(ocean))
            .sortWith(_ > _)
            .take(3)
            .reduce(_ * _)

    val ocean = Source
        .fromFile("input.txt")
        .getLines
        .map(s => s.toCharArray().map(_.toString).map(_.toInt).toVector)
        .toVector

    println(s"First part: ${firstPart(ocean)}")
    println(s"Second part: ${secondPart(ocean)}")

File: 2021/archived/Day09/Day9.fs
module Day9

open System.IO

let (<.>) f x y = f y x

let (|.|) (t: 'a array array) (x, y) = t.[y].[x]

let (<+>) e s = Set.add e s

let X table = table |> Array.head |> Array.length |> ((-) <.> 1)

let Y table = table |> Array.length |> ((-) <.> 1)

let neighbors X Y cell =
    let move (x, y) (vx, vy) = x + vx, y + vy
    let inBound (x, y) = not <| (x < 0 || x > X || y < 0 || y > Y)

    [| (0, 1); (0, -1); (1, 0); (-1, 0) |]
    |> Array.map (move cell)
    |> Array.filter inBound

let isLow table cell =
    let X = X table
    let Y = Y table
    let current = table|.|cell

    neighbors X Y cell
    |> Array.forall (fun cell -> table|.|cell > current)

let lowest table =
    let X = X table
    let Y = Y table

    [| for x in 0..X do
        for y in 0..Y -> (x, y) |]
    |> Array.filter (isLow table)

let firstPart table =
    lowest table
    |> Array.map (fun cell -> table|.|cell |> (+) 1)
    |> Array.sum

let basin table cell =
    let X = X table
    let Y = Y table

    let rec loop visted cell =
        if Set.contains cell visted || table|.|cell = 9 then
            visted
        else
            neighbors X Y cell
            |> Array.fold loop (cell <+> visted)

    loop Set.empty cell |> Set.count

let secondPart table =
    lowest table
    |> Array.map (basin table)
    |> Array.sortDescending
    |> Array.take 3
    |> Array.reduce (*)

let input = File.ReadAllLines "input.txt"
            |> Array.map Seq.toArray
            |> Array.map (Array.map (string >> int))

firstPart input |> printfn "First part: %d"
secondPart input |> printfn "Second part: %d"
